Do not update hasOutlineAutoAncestor when moving out a spanner's renderer

jace0x21 opened the pull request:
#### 25d4c61111c5948d715b1d1b51ba086c274d1d61
<pre>
Do not update hasOutlineAutoAncestor when moving out a spanner&apos;s renderer
<a href="https://bugs.webkit.org/show_bug.cgi?id=255956">https://bugs.webkit.org/show_bug.cgi?id=255956</a>
rdar://107953410

Reviewed by Alan Baradlay.

When we move out a spanner&apos;s renderer, we might set the renderer as having
a hasOutlineAutoAncestor value of true and leave the spanner placeholder
with a hasOutlineAutoAncestor value of false. This is because when we move
out the spanner&apos;s renderer we set hasOutlineAutoAncestor based on its new
parent. However we want to maintain a value consistent with spanner&apos;s placholder.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):
We want to check if the originalRenderer needs a repaint, not the placeholder.

* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::attachToRenderElementInternal):

Canonical link: <a href="https://commits.webkit.org/263501@main">https://commits.webkit.org/263501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef865c1d8f3bc7bf1fb1d967bc4d773673a5221c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6110 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4772 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4608 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5011 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4686 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4779 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4138 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6120 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2266 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4130 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9118 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4127 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5748 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4594 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3738 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4123 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1178 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8166 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4484 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
```diff
@@ -457,7 +457,8 @@ void RenderTreeBuilder::attachToRenderElementInternal(RenderElement& parent, Ren
 
     if (parent.hasOutlineAutoAncestor() || parent.outlineStyleForRepaint().outlineStyleIsAuto() == OutlineIsAuto::On)
-        newChild->setHasOutlineAutoAncestor();
+        if (!is<RenderMultiColumnSet>(newChild->previousSibling())) 
+            newChild->setHasOutlineAutoAncestor();
```
is this because in case of "being moved out" we use the incorrect parent to decide whether this bit should be set?

jace0x21 replied:
We decide whether this bit should be set based on the new parent it is being moved to which I don't think is the correct thing to do when moving out a spanner

alanbaradlay replied:
could you clarify that in the commit message, please?

jace0x21 committed:
Do not update hasOutlineAutoAncestor when moving out a spanner's renderer
https://bugs.webkit.org/show_bug.cgi?id=255956
rdar://107953410

Reviewed by Alan Baradlay.

When we move out a spanner's renderer, we might set the renderer as having
a hasOutlineAutoAncestor value of true and leave the spanner placeholder
with a hasOutlineAutoAncestor value of false. This is because when we move
out the spanner's renderer we set hasOutlineAutoAncestor based on its new
parent. However we want to maintain a value consistent with spanner's placholder.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):
We want to check if the originalRenderer needs a repaint, not the placeholder.

* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::attachToRenderElementInternal):

Canonical link: https://commits.webkit.org/263501@main
